Add ufbx importer article

lyuma opened the pull request:
The ufbx FBX importer is a big shift in import format support, and there are some gotchas and new features that users should know about.

This is still a draft for now since it does not yet have pictures.

lyuma committed:
Add ufbx importer article

The ufbx FBX importer is a big shift in import format support, and there are some gotchas and new features that users should know about.

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+date: 2099-09-09 00:00:00
+---
+
+Thanks to [a game-changing contribution by Samuli Raivio (bqqbarbhg) and Ernest Lee (fire)](https://github.com/godotengine/godot/pull/81746), Godot 4.3 now includes native .fbx support. This was made possible using Samuli’s free and open source [ufbx single-file fbx library(https://github.com/ufbx/ufbx). The ufbx importer will be used by default for all newly imported fbx files in Godot 4.3
```
```suggestion
Thanks to [a game-changing contribution by Samuli Raivio (bqqbarbhg) and Ernest Lee (fire)](https://github.com/godotengine/godot/pull/81746), Godot 4.3 now includes native .fbx support. This was made possible using Samuli’s free and open source [ufbx single-file .fbx library](https://github.com/ufbx/ufbx). The ufbx importer will be used by default for all newly imported .fbx files in Godot 4.3.
```

JohnVeness replied:
Also, "game-changing" is momentarily confusing, given this is a game engine! Maybe "a massive contribution" or "a significant contribution" or "a very welcome contribution" or similar...

lyuma replied:
I thought it would be humorous, but I suppose this pun only works so many times in the Game industry before it gets old ;-) I'll change it to the last one.

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+The new importer also contains many technical improvements and changes:
+* Imported scenes no longer have an extraneous "RootNode" top-level node in the scene.
+* Unit conversion (such as centimeters to meters)  is now applied to the mesh itself rather than assigned to the node scale of top-level nodes.
+* Improved support for PBR materials and lighting accuracy, within what the .fbx format supports. (.fbx tends to have inconsistent material support between applications)
```
```suggestion
* Improved support for PBR materials and lighting accuracy, within what the .fbx format supports. (.fbx tends to have inconsistent material support between applications).
```

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+* Support for geometric pivots. Pivot offsets are applied automatically where possible.
+* The option to create geometric helper nodes helps support pivots and inheritance modes not otherwise possible with traditional translation, rotation and scaling node transforms.
+* Automatic trimming of animations based on the start and end timestamp encoded in the .fbx file header.
+* Preserves the original material order from the DCC, rather than sorted by face index.
```
Might be worth defining DCC here?

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+
+# Improved import of animated characters
+
+Those who have worked with .fbx in the past may have encountered a plethora of issues when importing skeletons and animations, to the extent that it became common practice to import such animation files in Blender and re-export as glTF 2.0. WIth Godot 4.3, we hope that this will be a thing of the past. We have made the following improvements to the skeleton importer:
```
```suggestion
Those who have worked with .fbx in the past may have encountered a plethora of issues when importing skeletons and animations, to the extent that it became common practice to import such animation files in Blender and re-export as glTF 2.0. With Godot 4.3, we hope that this will be a thing of the past. We have made the following improvements to the skeleton importer:
```

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+* [#88824 (Retargeting option to use a template for silhouette)](https://github.com/godotengine/godot/pull/88824): This feature can be used to reference the RESET animation of a known good (T-pose) reference file. Having a separate reference pose .fbx is a common practice in some animation sets.
+* [#88821 (Allow preserving the initial bone pose in rest fixer)](https://github.com/godotengine/godot/pull/88821): This allows for importing a retargeted skeleton while keeping it in its initial pose.
+* [#88819 (Add new scene import option to import as Skeleton)](https://github.com/godotengine/godot/pull/88819): This solves cases especially common in .fbx files without any meshes, where animated skeletons import as a series of nodes.
+* Improvements to bone detection [#90050 (Add tips detection to auto mapping in bone mapper)](https://github.com/godotengine/godot/pull/90050): Improves support for Mixamo and ReadyPlayerMe finger bone conventions
```
```suggestion
* Improvements to bone detection [#90050 (Add tips detection to auto mapping in bone mapper)](https://github.com/godotengine/godot/pull/90050): Improves support for Mixamo and ReadyPlayerMe finger bone conventions.
```

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,55 @@
+
+# Issues and testing
+
+Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete. (This level of rigorous design and testing is a stark contrast to previous .fbx importer rewrites some of you may recall from Godot 3. We can now be confident when we say that the importer is a ...
```
```suggestion
Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete. This level of rigorous design and testing is a stark contrast to previous .fbx importer rewrites some of you may recall from Godot 3.
```
Removed last sentence as it seemed redundant (you've already just said it's feature-complete). Also, I didn't think the brackets were necessary.

JohnVeness commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,57 @@
+date: 2099-09-09 00:00:00
+---
+
+Thanks to [a very welcome contribution by Samuli Raivio (bqqbarbhg) and Ernest Lee (fire)](https://github.com/godotengine/godot/pull/81746), Godot 4.3 now includes native .fbx support. This was made possible using Samuli’s free and open source [ufbx single-file fbx library(https://github.com/ufbx/ufbx). The ufbx importer will be used by default for all newly imported fbx files in Godot 4.3
```
```suggestion
Thanks to [a very welcome contribution by Samuli Raivio (bqqbarbhg) and Ernest Lee (fire)](https://github.com/godotengine/godot/pull/81746), Godot 4.3 now includes native .fbx support. This was made possible using Samuli’s free and open source [ufbx single-file .fbx library](https://github.com/ufbx/ufbx). The ufbx importer will be used by default for all newly imported .fbx files in Godot 4.3.
```

fire commented:
Feedback from reviewers:

> Use extreme caution when changing the importer of an existing .fbx from FBX2glTF to ufbx: the two will have different node hierarchies, which will break modifications made in instanced scenes, NodePath references in scripts, affect rest poses of skeletons and possibly mix up surface override materials due to the change in surface ordering.

Was suggested to emphasize this more. Our current users are our most important users and the primary readers of the blog. Not everyone will go to the bottom of the page.

fire commented:
I would like some note of this feature https://github.com/godotengine/godot/issues/90314#issuecomment-2044107516 since our users mostly are from blender fbx but we are biasing towards the fbx sdk.

clayjohn commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,57 @@
+The new importer also contains many technical improvements and changes:
+* Tested against content created in [a wide range](https://github.com/ufbx/ufbx/tree/master/data) of professional 3D DCC (digital content creation) tools
+* Imported scenes no longer have an extraneous "RootNode" top-level node in the scene.
+* Unit conversion (such as centimeters to meters)  is now applied to the mesh itself rather than assigned to the node scale of top-level nodes.
```
```suggestion
* Unit conversion (such as centimeters to meters) is now applied to the mesh itself rather than assigned to the node scale of top-level nodes.
```

clayjohn commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,57 @@
+
+The ufbx importer is a major step in improving the accuracy and ease of use of Godot’s 3D importer. It is but one of many improvements, and there is more to come in future 4.x releases.
+
+There are sure to be many applications for the native .fbx importer. Our hope is it will improve interoperability with a wide variety of animation tools and libraries out in the wild. For example, I and the V-Sekai team has integrated ufbx into Unidot Importer to allow for interoperability with asset packs in .fbx format.
```
```suggestion
There are sure to be many applications for the native .fbx importer. Our hope is it will improve interoperability with a wide variety of animation tools and libraries out in the wild. For example, I and the V-Sekai team have integrated ufbx into Unidot Importer to allow for interoperability with asset packs in .fbx format.
```

clayjohn commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,57 @@
+
+# Issues and testing
+
+Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete. This level of rigorous design and testing is a stark contrast to previous .fbx importer rewrites some of you may recall from Godot 3.
```
```suggestion
Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete. This level of rigorous design and testing should result in a much better experience for .fbx users.
```

The current phrasing feels a bit disrespectful to the people who have worked on the current FBX import support in 3.x. I think its better to just focus on the positives as it sends the same message across

clayjohn reviewed (commented):
Looks great! I left a few suggestions for you to pick up once you finalize the post

fire commented on collections/_article/introducing-the-improved-ufbx-importer-in-godot-4-3.md:
```diff
@@ -0,0 +1,57 @@
+
+# Issues and testing
+
+Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete. This level of rigorous design and testing is a stark contrast to previous .fbx importer rewrites some of you may recall from Godot 3.
```
```suggestion
Samuli has spent exceptional time and effort testing the ufbx library on tens of thousands of .fbx test cases from a variety of sources, and we are proud to consider the new ufbx importer feature-complete.
```

lyuma replied:
replaced "with a high level of certainty"

lyuma commented:
Updated the text of the article

> I would like some note of this feature https://github.com/godotengine/godot/issues/90314#issuecomment-2044107516 since our users mostly are from blender fbx

That issue relates to FBX units scale and how to export from Blender with 1 meter units, so I added a note to address this:

> Another slight difference exists in the way FBX units are applied. The ufbx integration in Godot scales meshes directly rather than scaling nodes. Note that Blender by default exports objects at 100x scale, using 1cm units, To avoid unexpectedly scaled transforms, we recommend exporting from Blender with the setting "Apply Scalings" set to "FBX Units Scale", which will export the file as 1m units.

fire commented:
We probably want a image! `/storage/blog/covers/PlaceHolder.webp`

lyuma commented:
Added a nice cover image. Thanks, @adamscott 
I think this should make it ready to publish.

coppolaemilio commented:
Only thing that might be missing is Lyuma's profile picture, and then maybe you could link to v-sekai and unidot at the end when you mention them. Not a blocker, so up to you @lyuma

coppolaemilio committed:
Prepare blogpost to merge

coppolaemilio commented:
Thank you everyone! 💖

fire commented:
![image](https://github.com/user-attachments/assets/9758150c-11df-42d5-a837-fb8c12c5f648)

Link [report the issues you find](https://godotengine.org/godotengine/godot/issues) goes to `https://godotengine.org/godotengine/godot/issues` which is a dead link
